documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md

```diff
@@ -0,0 +1,277 @@
+within them changes significantly and the `script setup` simplifies the
+component. The decision wether to convert components to `script setup` would be
+at the discretion of the PR implementer.
+
+#### Testing
+
+Vue 3 recommends to use Vitest for better performance and integration with
+`vite` (which will replace `webpack` in our setup). I have not had an
+opportunity to test it out, but I think it would make sense to convert our unit
+tests to `Vitest`. The decision on whether to keep using the current
+dependencies or convert to `Vitest` should be based on the ease of conversion.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+Openverse frontend uses Nuxt 3, and all current features (including i18n,
+sitemaps, SEO) work.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+This section describes the known steps that need to be taken to migrate to
+Nuxt 3. However, it is possible that there will be some unknown steps that will
+be discovered during the implementation.
+
+- [ ] Add changes that can be done in Nuxt 2 (see "Changes to main")
+- [ ] Create a `nuxt3` branch with the first PR as described in the "Nuxt 3
+      branch"
+- [ ] Add all other Nuxt 3 changes, as described in the "Nuxt 3 branch"
+- [ ] Re-enable Playwright tests
+- [ ] Make all unit tests pass
+
+### Changes to main
+
+- Deprecation of CJS modules (the ones that use `require` and `module.exports`):
+  - `case` package that we use for `kebab-case`, `title` case and `camelCase`
+    function is CJS. We can either find replacements that use `ESM` modules,
+    vendor in these functions, or use a workaround that `vite` suggests (instead
+    of `const { kebab } = require "case"` use
+    `import casePkg; const { kebab } = casePkg;`)
+  - home gallery uses `require` when importing the images dynamically. There is
+    a small number of them, so we should just move them to `static` folder (or
+    `public` in Nuxt 3) and use static `import`s.
+  - [ ] #2342 Tailwind config should be converted to TypeScript.
+- Moving functionality inside the `setup` function would make it easier to
+  handle with Vue 3/Nuxt 3:
+  - Refactor `search.vue` page: move the route watcher and `fetchMedia`
+    functions to `setup`
+- Refactor image and audio single result pages to show the skeleton page when
+  the "image"/"audio" is `null`. This is required because the way data is
+  fetched changes, and _is_ the correct way it _should_ be set up (instead of
+  not handling the case of `image` being `null` and instead relying on the error
+  page redirect)
+- [x] #2234 Replace `onBeforeRouteEnter` router guards that don't have access to
+      `pinia` stores with router `middleware` that does
+- [x] #2233 Remove `@nuxtjs/style-resources` - this module was used to
+      auto-import `scss` variables into all components. We don't use `scss`
+      anymore
+
+### Changes in Nuxt 3 branch
```

Comment:
We need a step in this to remove the `nuxt-templates-overrides` too, right?

Reply:
Right!